project.clj

```diff
@@ -1,3 +1,58 @@
+(require 'clojure.java.io)
+
+(import '[java.util.zip ZipInputStream]
+        '[java.io FileOutputStream])
+
+(defmacro while-let [[sym expr] & body]
+  `(loop [~sym ~expr]
+     (when ~sym
+       ~@body
+       (recur ~expr))))
+
+(defn jdk-find [f]
+  (let [home (clojure.java.io/file (System/getProperty "java.home"))
+        parent (.getParentFile home)
+        paths [(clojure.java.io/file home f)
+               (clojure.java.io/file home "lib" f)
+               (clojure.java.io/file parent f)
+               (clojure.java.io/file parent "lib" f)]]
+    (->> paths (filter #(.canRead ^java.io.File %)) first str)))
+
+(def jdk-sources
+  (let [java-path->zip-path (fn [path]
+                   (some-> (clojure.java.io/resource path)
+                           ^java.net.JarURLConnection (. openConnection)
+                                      (. getJarFileURL)
+                                      clojure.java.io/as-file
+                                      str))]
+    (or (java-path->zip-path "java.base/java/lang/Object.java") ; JDK9+
+                   (java-path->zip-path "java/lang/Object.java") ; JDK8-
+        (jdk-find "src.zip"))))
+
+(defn uncompress [path target]
```

Comment:
For some reason simply adding `src.zip` to the classpath didn't work. I tried different variations (prepending `file:`, appending `!/`), no luck.

One has to uncompress src.zip instead, and add the resulting dir to the classpath.

(this adds ~5s to the build, doesn't seem worth caching... I'd be afraid of a given JDK creating a cache for a different JDK)

Reply:
I think that's fine.